rfcs/2022-03-15-11851-ingest-opentelemtry-traces.md

```diff
@@ -0,0 +1,368 @@
+# RFC 11851 - 2022-03-15 - Opentelemetry traces source
+
+This RFC aims to describes how to add an OpenTelemetry traces source to Vector and also address Vector internals
+adjustement required for future extension to other trace types.
+
+## Context
+
+- `datadog_agent` source supports receiving traces from the Datadog `trace-agent`
+- `datadog_traces` sink supports emitting traces to Datadog
+- OpenTelemetry traces are already supported by Datadog:
+  - Either with the [Datadog exporter][otlp-dd-exporter] using the Opentelemetry collector (without the `trace-agent`)
+  - Or with the `trace-agent` [configured to receive OpenTelemtry traces][otlp-traces-with-dd-agent] (both grpc and http
+    transport layer are supported
+
+### Usecases
+
+As the whole traces processing inside Vector is pretty new, documenting confirmed and most credible use cases in the
+near future will help to ensure changes will be implemented so they will be really useful to potential users. This also
+help to build something flexible enought to acomodate future needs.
+
+One identified scenario is to demux a trace flow based on some conditions that could be evaluated against any metadata
+for a single trace, a group of traces or per spans. From a config perspective this would expect to be functional with
+the following configuration:
+
+```yaml
+[...]
+sources:
+  otlp_traces:
+    type: opentelemetry_traces
+    address: "[::]:8081"
+    mode: grpc
+
+transforms:
+  set_key:
+    type: remap
+      source: |
+        if exists!(.tags.user_id) {
+          return
+        }
+        key = get_enrichment_table_record!("api_keys", { "user": .tags.user_id })
+        set_dd_api_key(key) # this does not exists yet
+      inputs:
+        - otlp_traces
+
+sinks:
+  dd_trace:
+    type: datadog_traces
+    default_api_key: 12345678abcdef
+    inputs:
+      - set_key
+```
+
+This demux/conditional action can be seen as an extension of what currently exists in Vector. Other kind of conditional
+action like the `filter` transform to discard traces base on certain metadata can be considered to be very similar, as
+this also involve evaluating a VRL condition on traces. The key problem here is to exposes traces and spans field in a
+way that the user can still manipulate those easily.
+
+This however raises the case of the granularity of a single event ; for instance multiple traces can bundles into a
+single payload in both OpenTelemetry and Datadog wire format. Enabling clear processing withou ambiguity advocate for a
+clear constraint that should be enforced by all future traces sources : **a single Vector event shall not hold data
+relative to more that one trace**.
```

Comment:
It's not clear to me how we will handle this?

If a payload contains multiple traces does the source just split this up into multiple events? Is there any requirement that the traces are together in the same payload when sent out from the sink? If so, that could get complicated, if not then we can probably relax!

Reply:
AFAIK there is no requirements send traces that would have been ingested together in the same batch from a sink perspective. I'll make that explicit that there is no reason to keep traces together.